File: run_analysis.R
# This script is the result of the final assignment for the Getting and Cleaning Data - Coursera course
# Please read additional details in CodeBook.md

# Important: This script assumes that data is downloaded and in the working directory.

# 1. Loading all the data files needed
features <- features <- gsub('(-|\\(\\)|,)', '', toupper(t(read.table('UCI HAR Dataset/features.txt')[2])))   # + Removing index column and cleaning sysmbols from feature names
y_train <- read.table('UCI HAR Dataset/train/y_train.txt')
y_test <- read.table('UCI HAR Dataset/test/y_test.txt')
x_train <- read.table('UCI HAR Dataset/train/X_train.txt')
x_test <- read.table('UCI HAR Dataset/test/X_test.txt')
activity_labels<- read.table('UCI HAR Dataset/activity_labels.txt')
subject_test <- read.table('UCI HAR Dataset/test/subject_test.txt')
subject_train <- read.table('UCI HAR Dataset/train/subject_train.txt')

# 2. Merging train and test data
y_data <- rbind(y_train, y_test)
x_data <- rbind(x_train, x_test)
subject_data <- rbind(subject_train, subject_test)

# 3. labeling y
y_data$rowid <- 1:nrow(y_data) # Adding row id because merging destroy the original order of rows
y_data <- merge(y_data, activity_labels, by="V1")
y_data <- y_data[order(y_data$rowid), ]
y_data <- y_data[, "V2"]

# 4. Filtering out features that aren't needed and naming the columns
filtered_features <- grep("MEAN|STD", features)
x_data <- x_data[, filtered_features]
names(x_data) <- features[filtered_features]

# 5. Binding x and y together
data <- cbind(x_data, y_data, subject_data)
colnames(data)[which(names(data) == "y_data")] <- "ACTIVITY"
colnames(data)[which(names(data) == "V1")] <- "SUBJECT"

# 6. save data: tidy_data.txt
write.table(data, file="tidy_data.txt", row.names = FALSE)

# 7. create averages and save data: tidy_data_means.txt
library(data.table)
dataTable <- data.table(data)
output <- dataTable[, lapply(.SD, mean), by=c("SUBJECT", "ACTIVITY")]
write.table(output, file="tidy_data_means.txt", row.names = FALSE)

File: run_analysis.py
# -*- coding: utf-8 -*-
"""
Created on Thu Feb 25 09:42:24 2016

@author: Janos_Strasszer
"""
import pandas as pd

DATA_DIR = 'c:/Development/PROJECTS/R/GettingAndCleaningData_Coursera/UCI HAR Dataset/'


# 1. Loading all the data files needed
features = pd.read_csv(DATA_DIR+'features.txt', header=None, sep=' ',
                       names=['index', 'feature_name'])
y_train = pd.read_csv(DATA_DIR+'train/y_train.txt', header=None, sep=' ', names=['activity_id'])
y_test = pd.read_csv(DATA_DIR+'test/y_test.txt', header=None, sep=' ', names=['activity_id'])
x_train = pd.read_csv(DATA_DIR+'train/X_train.txt', header=None, delimiter=r'\s+')
x_train.columns = features['feature_name'].values
x_test = pd.read_csv(DATA_DIR+'test/X_test.txt', header=None, delimiter=r'\s+')
x_test.columns = features['feature_name'].values
activity_labels = pd.read_csv(DATA_DIR+'activity_labels.txt', header=None, sep=' ',
                              names=["activity_id", "activity"])
subject_test = pd.read_csv(DATA_DIR+'test/subject_test.txt', header=None, names=["subject"])
subject_train = pd.read_csv(DATA_DIR+'train/subject_train.txt', header=None, names=["subject"])

# 2. Filtering out features that aren't needed
def filter_features(data_to_filter):
    f = list()
    for i in list(data_to_filter.columns):
        if "-std" in i.lower() or ("-mean" in i.lower()):
            f.append(i)
    return data_to_filter[f]

x_train=filter_features(x_train)
x_test=filter_features(x_test)

# 3. Binding x and y together and also subject
x_train['activity_id'] = y_train['activity_id']
x_train['subject'] = subject_train['subject']
x_test['activity_id'] = y_test['activity_id']
x_test['subject'] = subject_test['subject']


# 4. Merging train and test data
data = x_train.append(x_test)

# 4. labeling output column
data = pd.merge(data, activity_labels, on='activity_id', sort=False).drop('activity_id', axis=1)


# 6. save data
data.to_csv(DATA_DIR+'../tidy_data_py.txt', index=False, sep=' ', float_format='%.9f')
output = data.groupby(['subject', 'activity'], as_index=False).mean()
output.to_csv(DATA_DIR+'../tidy_data_means_py.txt', index=False, sep=' ', float_format='%.16f')
